python_modules/libraries/dagster-databricks/dagster_databricks/ext.py

```diff
@@ -0,0 +1,156 @@
+            message_reader=message_reader or ExtDbfsMessageReader(client=self.client),
+        ) as ext_context:
+            submit_task_dict = task.as_dict()
+            submit_task_dict["new_cluster"]["spark_env_vars"] = {
+                **submit_task_dict["new_cluster"].get("spark_env_vars", {}),
+                **(self.env or {}),
+                **ext_context.get_external_process_env_vars(),
+            }
+            task = jobs.SubmitTask.from_dict(submit_task_dict)
+            run_id = self.client.jobs.submit(
+                tasks=[task],
+                **(submit_args or {}),
+            ).bind()["run_id"]
+
+            while True:
+                run = self.client.jobs.get_run(run_id)
+                if run.state.life_cycle_state in (
+                    jobs.RunLifeCycleState.TERMINATED,
+                    jobs.RunLifeCycleState.SKIPPED,
+                ):
+                    if run.state.result_state == jobs.RunResultState.SUCCESS:
+                        return
+                    else:
+                        raise DagsterExternalExecutionError(
+                            f"Error running Databricks job: {run.state.state_message}"
+                        )
+                elif run.state.life_cycle_state == jobs.RunLifeCycleState.INTERNAL_ERROR:
+                    raise DagsterExternalExecutionError(
+                        f"Error running Databricks job: {run.state.state_message}"
+                    )
+                time.sleep(5)
+
+
+ExtDatabricks = ResourceParam[_ExtDatabricks]
+
+_CONTEXT_FILENAME = "context.json"
+
+
+@contextmanager
+def dbfs_tempdir(dbfs_client: files.DbfsAPI) -> Iterator[str]:
+    dirname = "".join(random.choices(string.ascii_letters, k=30))
+    tempdir = f"/tmp/{dirname}"
+    dbfs_client.mkdirs(tempdir)
+    try:
+        yield tempdir
+    finally:
+        dbfs_client.delete(tempdir, recursive=True)
+
+
+class ExtDbfsContextInjector(ExtContextInjector):
+    def __init__(self, *, client: WorkspaceClient):
+        super().__init__()
+        self.dbfs_client = files.DbfsAPI(client.api_client)
+
+    @contextmanager
+    def inject_context(self, context: "ExtContextData") -> Iterator[ExtParams]:
+        with dbfs_tempdir(self.dbfs_client) as tempdir:
+            path = os.path.join(tempdir, _CONTEXT_FILENAME)
+            contents = base64.b64encode(json.dumps(context).encode("utf-8")).decode("utf-8")
+            self.dbfs_client.put(path, contents=contents, overwrite=True)
+            yield {"path": path}
+
+
+class ExtDbfsMessageReader(ExtBlobStoreMessageReader):
+    tempdir: Optional[str] = None
+
+    def __init__(self, *, interval: int = 10, client: WorkspaceClient):
+        super().__init__(interval=interval)
+        self.dbfs_client = files.DbfsAPI(client.api_client)
+
+    @contextmanager
+    def setup(self) -> Iterator[None]:
+        with dbfs_tempdir(self.dbfs_client) as tempdir:
+            self.tempdir = tempdir
+            yield
+
```

Comment:
i thought we were doing this sort of mutable state business

Reply:
It was necessary given the existing setup of `ExtBlobStoreMessageReader` Updated with some changes to make it unnecessary.